WIP remove conntrack userspace dependency

aojea opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug

/kind cleanup
/kind deprecation

> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:

using netlink instead of the conntrack binary removes
the external dependency in the conntrack binary.

I have a follow up PR to improve the conntrack e2e coverage

Signed-off-by: Antonio Ojea <antonio.ojea.garcia@gmail.com>

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #91216

**Special notes for your reviewer**:

I've tried to add the new conntrack filter to the upstream netlink library
https://github.com/vishvananda/netlink/pull/544
 but it make sense to have it implemented in our repo 
so we can have a better control, in case we want to
fix or add new functionalities quickly.
Anyway, if the netlink library PR is accepted is straight forward to 
switch between one and the other.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubernetes not longer require conntrack-tools installed in the host
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aojea commented:
sig-network
/assign @danwinship @thockin 
sig-code-org aka sig-cleaning :)
/cc @liggitt @dims 
sig-testing
/cc @BenTheElder

aojea commented:
upps , missing some dependencies on the proxies,
let me clean it up and split it in different commits

aojea committed:
replace conntrack binary by netlink calls

aojea committed:
replace conntrack dockershim hostportmanager

aojea committed:
add conntrack interface

aojea committed:
replace kube-proxy iptables conntrack

thockin reviewed (commented):
I don't completely object to this, if we think it's really a problem, but I do have one significant concern.

Specifically - e2e testing this is tricky at best.  This is a fairly large amount of code to do something that we already have a thoroughly debugged tool to do.  How can we get as much confidence in this as we have in `conntrack` ?

I didn't do a very complete review yet because I have no real context.

aojea committed:
migrate user-space kube-proxy to conntrack

aojea committed:
replace conntrack ipvs proxier

aojea committed:
remove kube-proxy exec parameter

aojea committed:
update bazel deps

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -293,36 +274,22 @@ func TestDeleteEndpointConnections(t *testing.T) {
+			ct.OrigDstPort = int(tc.svcPort)
+			ct.NatDstIP = endpointIP
+			if tc.simulatedErr != "" && tc.simulatedErr != conntracktest.NoConnectionToDelete {
+				log.Error(nil, tc.simulatedErr)
```
I didn't understand well this part of the test, who was to log the "unexpected" error?
I've added this "hack" to state that the test passes with this "hack"

aojea committed:
remove exec kubemark dependency

aojea commented:
> I don't completely object to this, if we think it's really a problem, but I do have one significant concern.
> 
> Specifically - e2e testing this is tricky at best. This is a fairly large amount of code to do something that we already have a thoroughly debugged tool to do. How can we get as much confidence in this as we have in `conntrack` ?
> 
> I didn't do a very complete review yet because I have no real context.

The context is that we found out that only one e2e test failed if the conntrack binary was not present in the host https://github.com/kubernetes/kubernetes/issues/91216#issuecomment-633710458 , and this test was added only some weeks ago. 

The other thing is that when investigating some flakes, I realized that UDP service tests are generally more flaky than the others, the main difference with the UDP test are that UDP is lossless and they kick the conntrack deletion, this is just a hunch, I do not have any evidence of this, maybe the system is dropping those packets under load 🤷 

Definitely e2e test is the key here, I'm working in another PR to implement e2e for those conntrack scenarios that must go before this PR, in order to compare `conntrack-tools` against this approach.

I'm not having a strong opinion on this, just wanted a quick feedback like the one you just gave. If the risk is high and the benefit low, we can discard it.

aojea commented on pkg/util/conntrack/conntrack_test.go:
```diff
@@ -17,301 +17,318 @@ limitations under the License.
-
-	for i := range testCases {
-		err := Exec(&fexec, testCases[i]...)
+func TestClearEntriesForIP(t *testing.T) {
```
This test require root privileges, otherwise it fails, is it possible to give it those privileges or should I mock it?
It will be great being able to test this with real flows though

BenTheElder replied:
We do run as root in CI currently but it's kinda not great for local developers ...

aojea replied:
it needs CAP_NET_ADMIN, maybe the pod that is running does not allow it despite being root?

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -293,36 +270,22 @@ func TestDeleteEndpointConnections(t *testing.T) {
+				klog.Errorf(tc.simulatedErr)
 			}
-			actualCommand := strings.Join(fcmd.CombinedOutputLog[fexec.CommandCalls-1], " ")
-			if actualCommand != expectCommand {
```
I have to come back to this test, I'm not getting exactly how this is testing if the conntrack gets deleted

aojea committed:
e2e conntrack tests

deflake current e2e test
"should be able to preserve UDP traffic when server pod cycles for a
NodePort service" and reorganize the code in the e2e framework

Signed-off-by: Antonio Ojea <antonio.ojea.garcia@gmail.com>

BenTheElder commented:
Can we just featuregate this and use the old codepath / exec by default?

Would be curious to see the test results first of course.

thockin commented:
I would definitely be AGAINST coding our own iptables lib.

This is somewhat less complicated, so I am not for sure against this, but
it seems that we need internalize details about netlink which increases our
cognitive burden.  The gain has to be worthwhile.

What is the alternative?  Fail harder in the absence of conntrack?

On Sat, Jun 13, 2020, 2:39 AM Antonio Ojea <notifications@github.com> wrote:

> I don't completely object to this, if we think it's really a problem, but
> I do have one significant concern.
>
> Specifically - e2e testing this is tricky at best. This is a fairly large
> amount of code to do something that we already have a thoroughly debugged
> tool to do. How can we get as much confidence in this as we have in
> conntrack ?
>
> I didn't do a very complete review yet because I have no real context.
>
> The context is that we found out that only one e2e test failed if the
> conntrack binary was not present in the host #91216 (comment)
> <https://github.com/kubernetes/kubernetes/issues/91216#issuecomment-633710458>
> , and this test was added only some weeks ago.
>
> The other thing is that when investigating some flakes, I realized that
> UDP service tests are generally more flaky than the others, the main
> difference with the UDP test are that UDP is lossless and they kick the
> conntrack deletion, this is just a hunch, I do not have any evidence of
> this, maybe the system is dropping those packets under load 🤷
>
> Definitely e2e test is the key here, I'm working in another PR to
> implement e2e for those conntrack scenarios that must go before this PR, in
> order to compare conntrack-tools against this approach.
>
> I'm not having a strong opinion on this, just wanted a quick feedback like
> the one you just gave. If the risk is high and the benefit low, we can
> discard it.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/92076#issuecomment-643598753>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVBCZOYWOBYNKEIGBUTRWNCOZANCNFSM4N4OJK3Q>
> .
>

BenTheElder commented:
> Fail harder in the absence of conntrack?

Discussion with @danwinship on a PR for exactly that is what led to this PR, FWIW :upside_down_face: 
https://github.com/kubernetes/kubernetes/pull/91486#discussion_r431817534

aojea commented:
> I would definitely be AGAINST coding our own iptables lib.

I'm totally with you on that 😄

aojea committed:
flush conntrack entries before iptables

aojea commented:
failed conntracks
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92076/pull-kubernetes-e2e-kind-ipv6/1272250397695676419
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92076/pull-kubernetes-e2e-gce-ubuntu-containerd/1272250397699870720/

aojea commented:
/retest

aojea commented:
:/ failed conntracks
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92076/pull-kubernetes-e2e-kind/1272269515035512836
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92076/pull-kubernetes-e2e-gce-ubuntu-containerd/1272269515035512835/

aojea commented:
this is the flakiest one by far :/

let's not spam people
/remove-assign @danwinship @thockin
/remove-cc @liggitt @dims
/remove-cc @BenTheElder

aojea commented:
/unassign @danwinship @thockin
/uncc @liggitt @dims
/uncc @BenTheElder

danwinship commented:
>  This is a fairly large amount of code to do something that we already have a thoroughly debugged tool to do.

Yeah... I saw that `vishvananda/netlink` had some conntrack-related code and was hoping that we could use that to replace our calls to` /sbin/conntrack` they way we had previously (in openshift-sdn and ovn-kubernetes) replaced all of our calls to `/sbin/ip` with netlink calls, which made everything much cleaner. But I guess the addressing/routing code in `vishvananda/netlink` is a little higher-level than the conntrack code.

> > Fail harder in the absence of conntrack?
> 
> Discussion with @danwinship on a PR for exactly that is what led to this PR, FWIW upside_down_face
> [#91486 (comment)](https://github.com/kubernetes/kubernetes/pull/91486#discussion_r431817534)

Yeah, I didn't mean to suggest that I was against that PR, just that it wasn't failing hard enough, and _also_ that another approach would be to rewrite it to not require an external binary.

Maybe we should go back to that for now, and maybe also keep the idea of the `ConntrackClearer` abstraction from this PR (just still implemented using exec on the backend), and then if Antonio's netlink PR to add the filter code there gets accepted, we can think about porting `ConntrackCleaner` to use it later?

thockin commented:
That sounds OK to me.  Making conntrack.Clearer or something seems like a
good standalone abstraction improvement

On Mon, Jun 15, 2020 at 2:31 PM Dan Winship <notifications@github.com>
wrote:

> This is a fairly large amount of code to do something that we already have
> a thoroughly debugged tool to do.
>
> Yeah... I saw that vishvananda/netlink had some conntrack-related code
> and was hoping that we could use that to replace our calls to
> /sbin/conntrack they way we had previously (in openshift-sdn and
> ovn-kubernetes) replaced all of our calls to /sbin/ip with netlink calls,
> which made everything much cleaner. But I guess the addressing/routing code
> in vishvananda/netlink is a little higher-level than the conntrack code.
>
> Fail harder in the absence of conntrack?
>
> Discussion with @danwinship <https://github.com/danwinship> on a PR for
> exactly that is what led to this PR, FWIW upside_down_face
> #91486 (comment)
> <https://github.com/kubernetes/kubernetes/pull/91486#discussion_r431817534>
>
> Yeah, I didn't mean to suggest that I was against that PR, just that it
> wasn't failing hard enough, and *also* that another approach would be to
> rewrite it to not require an external binary.
>
> Maybe we should go back to that for now, and maybe also keep the idea of
> the ConntrackClearer abstraction from this PR (just still implemented
> using exec on the backend), and then if Antonio's netlink PR to add the
> filter code there gets accepted, we can think about porting
> ConntrackCleaner to use it later?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/92076#issuecomment-644401143>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVARLQJBLIBV623UT6DRW2HJ5ANCNFSM4N4OJK3Q>
> .
>
